tendermint/src/lite/verifier.rs

```diff
@@ -154,3 +164,212 @@ where
+    let pivot_vals = req.validator_set(pivot)?;
+
+    verify(&pivot_header, &pivot_vals)?;
+
+    // Can we trust pivot header based on trusted_state?
+    can_trust_bisection(
+        trusted_state,
+        &pivot_header,
+        trust_threshold,
+        trusting_period,
+        now,
+        req,
+        store,
+    )?;
+    // Trust the header in between the trusted and (still) untrusted height:
+    let pivot_next_vals = req.validator_set(pivot + 1)?;
+    let pivot_trusted = TS::new(&pivot_header, &pivot_next_vals);
+    store.add(&pivot_trusted)?;
+
+    // Can we trust the (still) untrusted header based on the (now trusted) "pivot header"?
+    can_trust_bisection(
+        &pivot_trusted,
+        untrusted_header,
+        trust_threshold,
+        trusting_period,
+        now,
+        req,
+        store,
+    )?;
+    // Add header (and corresponding next validators) to trusted state store:
+    let untrusted_next_vals = req.validator_set(untrusted_header.header().height().increment())?;
+    let untrusted_state = TS::new(untrusted_header, &untrusted_next_vals);
+    store.add(&untrusted_state)?;
+
+    Ok(())
+}
+
+/// This function captures the high level logic of the light client verification, i.e.,
+/// an application call to the light client module to (optionally download) and
+/// verify a header for some height.
+pub fn verify_header<TS, SH, C, L, S, R>(
+    height: Height,
+    trust_threshold: &L,
+    trusting_period: &Duration,
+    now: &SystemTime,
+    req: &R,
+    store: &mut S,
+) -> Result<(), Error>
+where
+    TS: TrustedState<LastHeader = SH, ValidatorSet = C::ValidatorSet>,
+    L: TrustThreshold,
+    S: Store<TrustedState = TS>,
+    R: Requester<SignedHeader = SH, ValidatorSet = C::ValidatorSet>,
+    C: Commit,
+    SH: SignedHeader<Commit = C>,
+{
+    // Check if we already trusted a header at the given height and it didn't expire:
+    if let Ok(ts2) = store.get(height) {
+        is_within_trust_period(ts2.last_header().header(), trusting_period, now)?
+    }
+
+    // We haven't trusted a header at given height yet. Request it:
+    let sh2 = req.signed_header(height)?;
+    let sh2_vals = req.validator_set(height)?;
+    verify(&sh2, &sh2_vals)?;
+    is_within_trust_period(sh2.header(), trusting_period, now)?;
+
+    // Get the highest trusted header with height lower than sh2's.
+    let sh1_trusted = store.get_smaller_or_equal(height)?;
+    can_trust_bisection(
+        &sh1_trusted,
+        &sh2,
+        trust_threshold,
+        trusting_period,
+        now,
+        req,
+        store,
+    )?;
+
+    // TODO(Liamsi): The spec re-checks if we are still in trust period here
+    // and stores the now trusted header again.
+    // Figure out if we want to store it here or in can_trust_bisection!
+    // My understanding is: with the current impl, we either bubbled up an
+    // error, or, successfully added sh2 (and its nex vals) to the trusted store here.
```

Comment:
See: https://github.com/tendermint/spec/pull/61/files#r361808209
and https://github.com/tendermint/spec/pull/61/files#r361808447